bin/commandTopicConsumer.py

```diff
@@ -0,0 +1,165 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+#
+# Dump ksqlDB persistent queries to stdout.
+#
+# Usage:
+#  commandTopicConsumer.py -f <client-conf-file>
+#
+
+from confluent_kafka import KafkaError, KafkaException, version
+from confluent_kafka import Producer, Consumer
+import json
+import logging
+import argparse
+import uuid
+import sys
+import re
+
+
+class CommandRecord (object):
+    def __init__(self, stmt):
+        self.stmt = stmt
+
+    def __str__(self):
+        return "({})".format(self.stmt)
+
+    @classmethod
+    def deserialize(cls, binstr):
+        d = json.loads(binstr)
+        return CommandRecord(d['statement'])
+
+max_offset = -1001
+def consumer_latest_offsets(consumer, partitions):
+    global max_offset
+    for p in partitions:
+        high_water = consumer.get_watermark_offsets(p)[1]
+        if high_water >= max_offset:
+            max_offset = high_water
+
+class CommandConsumer(object):
+    def __init__(self, ksqlServiceId, conf):
+        self.consumer = Consumer(conf)
+        self.topic = '_confluent-ksql-{}_command_topic'.format(ksqlServiceId)
+
+    def consumer_run(self):
+        global max_offset
+        self.consumer.subscribe([self.topic], on_assign=consumer_latest_offsets)
+        #print("max offset = ", max_offset)
+        self.msg_cnt = 0
+        self.msg_err_cnt = 0
+        stmts = {}
+        try:
+            while True:
+                msg = self.consumer.poll(0.2)
+                if msg is None:
+                    continue
+
+                if msg.error() is not None:
+                    print("consumer: error: {}".format(msg.error()))
+                    self.consumer_err_cnt += 1
+                    continue
+
+                try:
+                    #print("Read msg with offset ", msg.offset())
+                    self.msg_cnt += 1
+                    record = CommandRecord.deserialize(msg.value())
+                    match = re.search(r'(?:create|drop) (?:stream|table) ([a-zA-z0-9-]+?)(:?\(|AS|\s|;)', record.stmt, re.I)
+                    if match:
+                        name = match.group(1).upper()
+                        if name == "KSQL_PROCESSING_LOG":
+                            continue
+                        if name not in stmts:
+                            stmts[name] = []
+                        stmts[name].append(record.stmt)
+                    else:
+                        match2 = re.search(r'(?:terminate) (?:ctas|csas)_(.+?)_', record.stmt, re.I)
+                        if match2:
+                            name = match2.group(1).upper()
+                            stmts[name].append(record.stmt)
+                        else:
+                            print("No match %s",record)
+                    # High watermark is +2 from last offset
+                    if msg.offset() >= max_offset-2:
```

Comment:
Are we sure it's +2, and not 'one past the last'?

Reply:
The max offset is 1 past he last one but the message start counting from 0.

```
max offset =  10
Read msg with offset  0
Read msg with offset  1
Read msg with offset  2
Read msg with offset  3
Read msg with offset  4
Read msg with offset  5
Read msg with offset  6
Read msg with offset  7
Read msg with offset  8
Consumed 9 messages, erroneous message = 0.
```

Reply:
It is +1, you right